### DISCUSSION TITLE
Installation on Ubuntu 20.04 - help

### DISCUSSION THREAD

**Discussion by Anto79-ops** [timmo001/system-bridge#1829] [Help] [answered]
Hi! 

Trying to get this to work on Ubuntu 20.04. First I ran these commands without any isses (actually, it said I already had them installed)

```
sudo apt update
sudo apt install lshw upower
sudo apt install python3 python3-pip python3-setuptools python3-wheel
```

then I ran this line:

`python -m pip install --upgrade systembridgeshared systembridgebackend systembridgecli systembridgefrontend systembridgegui
`
but got an error message saying python does not exist and suggested I use python3 from  command 'python3' from deb python3 so I did:

`python3 -m pip install --upgrade systembridgeshared systembridgebackend systembridgecli systembridgefrontend systembridgegui
`
it was working untill the end, where some read erros saying systembrigegui

```
Collecting pyperclip==1.8.2
  Using cached pyperclip-1.8.2.tar.gz (20 kB)
ERROR: Could not find a version that satisfies the requirement PySide6==6.3.1 (from systembridgegui) (from versions: 6.0.0a1.dev1606911628, 6.0.0, 6.0.1, 6.0.2, 6.0.3, 6.0.4, 6.1.0, 6.1.1, 6.1.2, 6.1.3, 6.2.0, 6.2.1, 6.2.2, 6.2.2.1, 6.2.3, 6.2.4)
ERROR: No matching distribution found for PySide6==6.3.1 (from systembridgegui)
```
at this point I realised I dont need the GUI, its headless ubuntu and I want this to feed Home Assistant on a Rpi so I went with the headless install:

`python3 -m pip install --upgrade systembridgeshared systembridgebackend systembridgecli
`
This installed without an issues.

Then I typed:

`xdg-user-dirs-update
`
and then this:

`python3 -m systembridgebackend --no-gui
`
However, it doesn't start and presented with this error:

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/antonio/.local/lib/python3.8/site-packages/systembridgebackend/__main__.py", line 6, in <module>
    from systembridgeshared.const import SETTING_AUTOSTART, SETTING_LOG_LEVEL
  File "/home/antonio/.local/lib/python3.8/site-packages/systembridgeshared/const.py", line 3, in <module>
    from systembridgeshared.models.bridge import Bridge
  File "/home/antonio/.local/lib/python3.8/site-packages/systembridgeshared/models/bridge.py", line 9, in <module>
    class Bridge(BaseModel):
  File "pydantic/main.py", line 188, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/typing.py", line 419, in pydantic.typing.resolve_annotations
    For example:
  File "/usr/lib/python3.8/typing.py", line 270, in _eval_type
    return t._evaluate(globalns, localns)
  File "/usr/lib/python3.8/typing.py", line 518, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
TypeError: 'type' object is not subscriptable

```

Any help to  get this running? thank you

**Comment by timmo001** [comment]
This comes from https://github.com/timmo001/system-bridge/issues/1804

**Comment by Anto79-ops** [comment]
Hi thanks!

Got it running once python 3.10 was used. 

So it's running. Home Assistant discovered is almost instantly, got the token and everything but, it doesn;t seem to be showing entities and HA reports timed out connection. The program is running using

`python -m systembridgebackend --no-gui
`
and things are flying up the screen, green, red, yellow, white...not sure if I'm supposed to configure settings?

thanks in advanced

**Reply by timmo001** [reply]
Have you updated to the latest version of Home Assistant? This was released yesterday. Any older version is incompatible with 3.x.x

**Comment by Anto79-ops** [comment]
Thanks for suggesting that.  I updated yesterday to the realese version (from beta). I now have these versions installed:

Home Assistant 2022.7.0
Supervisor 2022.07.0
Operating System 8.2
Frontend 20220706.0 - latest

Just to make sure, I deleted the integration this morning in HA and added back (it auto-discovered actually, soon as I started the service on my Ubuntu machine), and then placed my token in HA and it said sucessful.

Here's what i'm seeing, while the program is running below:

![system_bridge1](https://user-images.githubusercontent.com/79593761/177784761-cf99e4be-508f-40ea-b3a1-b6ef6247a341.jpg)

HA reports this:

![system_bridge2](https://user-images.githubusercontent.com/79593761/177784875-ae7d2939-eda3-4f9b-a59a-d17ea51eaacc.jpg)

Once it works, are the entities supposed to show up here or somewhere else?

Also, could you please tell me how I can check for the path of installation? I'm trying to create the systemmd service but I think I have my path wrong.

thanks again for your help

**Comment by Anto79-ops** [comment]
and sometimes the error message does dissappear and the tile looks like this:

![system_bridge3](https://user-images.githubusercontent.com/79593761/177786200-cc7630e0-d10b-4c83-ada9-c91d3bef966c.jpg)

**Comment by Anto79-ops** [comment]
update: I got the systemd working

still not seeing things reporting in HA just a blank tile (as above). 

Should there be enties reporting here or will it be in my developer tools entites? thanks

maybe related to this error in the HA Log?

```
Logger: homeassistant
Source: components/system_bridge/coordinator.py:113
First occurred: 3:28:37 PM (7 occurrences)
Last logged: 3:36:55 PM

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/system_bridge/coordinator.py", line 113, in _listen_for_data
    await self.websocket_client.listen(callback=self.async_handle_module)
  File "/usr/local/lib/python3.10/site-packages/systembridgeconnector/websocket_client.py", line 513, in listen
    await self.listen_for_messages(callback=_callback_message)
  File "/usr/local/lib/python3.10/site-packages/systembridgeconnector/websocket_client.py", line 528, in listen_for_messages
    await callback(message)
  File "/usr/local/lib/python3.10/site-packages/systembridgeconnector/websocket_client.py", line 466, in _callback_message
    model(**message[EVENT_DATA]),
  File "/usr/local/lib/python3.10/site-packages/pydantic/main.py", line 331, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for Cpu
temperature
  none is not an allowed value (type=type_error.none.not_allowed)

```

**Reply by timmo001** [reply]
That error will be what is blocking the integration. This was fixed in an update to `systembridgeconnector` recently. I will open up a PR on home-assistant/core to update this.

**Reply by Anto79-ops** [reply]
Thanks and thanks for the integration and upkeep! 
I'll work on getting this installed on a windows computer in the meantime.

**Comment by Anto79-ops** [comment]
Hi! I noticed there was an update to HA in 7.3 that address this issue above. That's awesome, thanks!

I've updated HA to the latest as of today. Just curious, do I also have to update the Ubuntu instace? If so, how do I actually update the Ubuntu instance of this integration? My install steps are above.

thanks

**Reply by timmo001** [reply]
For the above, the Home Assistant update was the fix as the model was updated in the connector

**Comment by Anto79-ops** [comment]
thanks! I tried to get it running ...by adding the intergration again...soon as I started it on my Ubunutu machine, it showed up immediately in HA, BUT for some reason, its now not accepting my api token.

I even created a new token(s):

`python3.10 -m systembridgecli api-key --reset`

and pasting it into the text box, i press submit and it waits a few seconds and say "(!)failed to connect"

**Comment by timmo001** [comment, marked as answer]
Please try the latest version. This allows uses from python 3.9 instead of newer versions.

https://github.com/timmo001/system-bridge/releases/tag/3.4.3

**Comment by Anto79-ops** [comment]
thanks/ just updated tp 3.4.3 using the --upgrade install command and all updated.

Its still not connecting to HA (but works using the browser), BUT it doesn't seem the PR on the HA side has been merged yet..

https://github.com/home-assistant/core/issues/75532#issuecomment-1190772091

**Reply by timmo001** [reply]
There is a pending fix in that issue. Hang tight!

**Comment by Anto79-ops** [comment]
looks like the fix is ready but still hasn't merged yet. Maybe it needs a smash bump :)

**Comment by Anto79-ops** [comment]
Hey @timmo001 

just updated to the latest HA core 2022.8.5, was excited to see they finally merged your changes. 

I started system bridge on Ubuntu 22.04, HA auto discovered, I put in my API token and it accepted it!

However, I'm still not seeing anything coming into HA. The system bridge tile has no entities OR goes to timed out (as above).

Some log errors on the HA side are show here:

```

Logger: systembridgeconnector.websocket_client.WebSocketClient
Source: components/system_bridge/coordinator.py:116
First occurred: 10:15:16 PM (3 occurrences)
Last logged: 10:21:58 PM

Future already set for response ID: b8f6a2129e0b4c7182c8c690eadbab82
Future already set for response ID: 070de91f82fe4a01bec0040a720b0436
Future already set for response ID: 4a041b52baf7480da4f6064b769573c8

```
and

```
Logger: homeassistant.config_entries
Source: config_entries.py:388
First occurred: 10:15:46 PM (1 occurrences)
Last logged: 10:15:46 PM

Config entry '192.168.1.139' for system_bridge integration not ready yet: Timed out waiting for 192.168.1.139 (192.168.1.139).; Retrying in background
```

Using the web browser to log in to system bridge, works and I see all the information from my Ubuntu machine.

**Comment by Anto79-ops** [comment]
oh, and another log error here, hope this helps



```
Logger: homeassistant
Source: /usr/src/homeassistant/homeassistant/runner.py:96
First occurred: 10:19:58 PM (1 occurrences)
Last logged: 10:19:58 PM

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/systembridgeconnector/websocket_client.py", line 492, in listen
    await self.listen_for_messages(callback=_callback_message)
  File "/usr/local/lib/python3.10/site-packages/systembridgeconnector/websocket_client.py", line 505, in listen_for_messages
    message = await self.receive_message()
  File "/usr/local/lib/python3.10/site-packages/systembridgeconnector/websocket_client.py", line 527, in receive_message
    raise ConnectionClosedException("Connection closed to server")
systembridgeconnector.exceptions.ConnectionClosedException: Connection closed to server
```
